sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,1512 @@
+            for idx in prange(self.n_X, schedule='static'):
+                simultaneous_sort(
+                    &self.argkmin_distances[idx, 0],
+                    &self.argkmin_indices[idx, 0],
+                    self.k,
+                )
+        return
+
+    cdef void compute_exact_distances(self) nogil:
+        cdef:
+            ITYPE_t i, j
+            ITYPE_t[:, ::1] Y_indices = self.argkmin_indices
+            DTYPE_t[:, ::1] distances = self.argkmin_distances
+        for i in prange(self.n_X, schedule='static', nogil=True,
+                        num_threads=self.effective_omp_n_thread):
+            for j in range(self.k):
+                distances[i, j] = self._datasets_pair.distance_metric._rdist_to_dist(
+                    # Guard against eventual -0., causing nan production.
+                    distances[i, j] if distances[i, j] > 0. else 0.
+                )
+
+    @final
+    def compute(self,
+       str strategy="auto",
+       bint return_distance=False,
+    ):
+        """Computes the reduction of vectors (rows) of X on Y.
+
+        Parameters
+        ----------
+        strategy: str, {'auto', 'parallel_on_X', 'parallel_on_Y'}, default='auto'
+            The chunking strategy defining which dataset parallelization are made on.
+
+            Strategies differs on the dispatching they use for chunks on threads:
+                 - 'parallel_on_X' dispatches chunks of X uniformly on threads.
+                 Each thread then iterates on all the chunks of Y. This strategy is
+                 embarrassingly parallel and comes with no datastructures synchronisation
+                 but is less used in practice (because X is smaller than Y generally).
+                 - 'parallel_on_Y' dispatches chunks of Y uniformly on threads.
+                 Each thread then iterates on all the chunks of X. This strategy is
+                 embarrassingly parallel but uses intermediate datastructures
+                 synchronisation. However it is more useful in practice (because Y is
+                 larger than X generally).
+                 -'auto' relies on a simple heuristic to choose between 'parallel_on_X'
+                 and 'parallel_on_Y'.
+
+        return_distance: boolean, default=False
+            Return distances between each X vector and its
+            argkmin if set to True.
+
+        Returns
+        -------
+        distances: ndarray of shape (n, k)
+            Distances between each X vector and its argkmin
+            in Y. Only returned if ``return_distance=True``.
+
+        indices: ndarray of shape (n, k)
+            Indices of argkmin of vectors of X in Y.
+        """
+
+        # Results returned by ArgKmin.compute used as the main heaps
+        self.argkmin_indices = np.full((self.n_X, self.k), 0, dtype=ITYPE)
+        self.argkmin_distances = np.full((self.n_X, self.k), DBL_MAX, dtype=DTYPE)
+
+        if strategy == 'auto':
```

Comment:
Maybe we could also look-up `sklearn.get_config().get("pairwise_dist_parallel_strategy")` to make it easier to measure the impact of this choice on the performance. If it's not `None` we use it. If `None`, use the heuristic.